### ISSUE TITLE
Type checker should handle simple cases of EAFP

### ISSUE THREAD

**Issue by justin-snyder-slgg** [microsoft/pyright#3720]
**Is your feature request related to a problem? Please describe.**

Given this snippet
```python
    def do_thing(self, items: Iterable):
        try:
            if len(items) >= self._max_len:
                ...  # some logic goes here
        except TypeError:
            pass
        ...
```

this triggers a `reportGeneralTypeIssues` on `len(items)` because Iterable has no `__len__`, which would be correct if it weren't for the `except TypeError`

**Describe the solution you'd like**

No error should be reported in this situation because the TypeError handles the failure of `len`. (Technically it also catches lots of other invalid arguments as well, so perhaps it is best if EAFP rules only apply to singular statements)

**Additional context**

There are plenty of other cases where EAFP applies to typechecking, e.g. `AttributeError`

This issue kind of straddles the line between feature and bug.

**Comment by erictraut** [comment]
A type checker doesn't have any information about what exception types are raised by specific calls or the circumstances around which such exceptions are raised.

Exceptions should not be used for normal code flow. As the name implies, exceptions are intended only for exceptional situations. If you are using exceptions for standard code flow, I recommend that you consider switching to normal conditional statements instead. If you want your code to be robust, fast, and work well with static type checking, EAFP is not a good approach.

**Comment by justin-snyder-slgg** [comment]
Adding a defensive check that the Python runtime would have done anyway doesn't make code faster. In other languages, yeah, sure, this would be an abuse of exceptions, but not so in Python, where it is commonplace to do things such as

```python
try:
    thing = int(some_string)
except ValueError:
    thing = 0
```

It's not exceptional to be given a string value that can't be converted to an integer, yet we use exceptions here.

In this specific case, a `Collection` (which supports `len`) is a more specific kind of `Iterable`, so this isn't a particularly crazy operation. I *can* LBYL here, true, but generally it's not out of the question in Python to try something and fall back if it doesn't work.

It's not the end of the world if this isn't understood by pyright. I don't base my entire python programming existence around appeasing type checkers.

**Comment by Azureblade3808** [comment]
Use `# type: ignore` where type-checks are expected to fail.
